core/modules/nfv_ctrl.cc

```diff
@@ -206,11 +207,166 @@ CommandResponse NFVCtrl::Init(const bess::pb::NFVCtrlArg &arg) {
     bess::utils::slo_ns = arg.slo_ns();
   }
 
-  return CommandSuccess();
+  if (!arg.port().empty()) {
+    const char *port_name = arg.port().c_str();
+    const auto &it = PortBuilder::all_ports().find(port_name);
+    if (it == PortBuilder::all_ports().end()) {
+      LOG(INFO) << "Failed to find port";
+      return CommandFailure(ENODEV, "Port %s not found", port_name);
+    }
+    port_ = ((PMDPort*)it->second);
+    //create and save the core to bucket mapping
+    for (uint16_t i = 0; i < total_core_count_; i++) {
+      core_bucket_mapping_[i] = std::vector<uint16_t>();
+    }
+    for(uint16_t i = 0; i < port_->reta_size_; i++) {
+      uint16_t core_id = port_->reta_table_[i];
+      assert(core_id < total_core_count_);
+      core_bucket_mapping_[core_id].push_back(i);
+    }
+  }
+  std::ifstream file("long_term_threshold");
+  while (!file.eof()) {
+    uint64_t pps;
+    uint64_t flow_count;
+    file >> flow_count;
+    file >> pps;
+    threshold_[flow_count] = pps;
+  }
+   
+   return CommandSuccess();
+}
+/*
+void WriteToGurobi(uint32_t num_cores, std::vector<float> flow_rates, float latency_bound) {
+  LOG(INFO) << num_cores << flow_rates.size() << latency_bound;
+  std::ofstream file_out;
+  file_out.open("./gurobi_in");
+  file_out << num_cores <<std::endl;
+  file_out << flow_rates.size() << std::endl;
+  file_out << std::fixed <<latency_bound <<std::endl;
+  for (auto &it : flow_rates) {
+    file_out << it<< std::endl;
+  }
+ file_out.close();
+}
+*/
+
+
+std::map<uint16_t, uint16_t> NFVCtrl::findMoves(double flow_rate_per_cpu[], std::vector<uint16_t> to_be_moved, std::vector<double> flow_rate_per_bucket) {
+  std::map<uint16_t, uint16_t> moves;
+  //if not found return error
+  for (auto it: to_be_moved) {
+    double flow_rate = flow_rate_per_bucket[it];
+    bool found = false;
+    for (uint16_t i = 0; i < total_core_count_; i++) {
+      if (flow_rate_per_cpu[i]>0 && flow_rate_per_cpu[i] + flow_rate < (threshold_[10000] * (1-HEAD_ROOM))) {
+        flow_rate_per_cpu[i] += flow_rate;
+        moves[it] = i;
+        core_bucket_mapping_[i].push_back(it);
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      // find a cpu with 0 flow rate and enable it.
+      for (uint16_t i = 0; i < total_core_count_; i++) {
+        if (flow_rate_per_cpu[i] == 0) {
+          flow_rate_per_cpu[i] += flow_rate;
+          moves[it] = i;
+          core_bucket_mapping_[i].push_back(it);
+          LOG(INFO) << "Added new CPU: " << i;
+        }
+      }
+    }
+  }
+  return moves;
+}
+
+std::map<uint16_t, uint16_t> NFVCtrl::longTermOptimization(std::vector<double> flow_rate_per_bucket) {
```

Comment:
When implementing an algorithm, please add a comment section to introduce all steps in the algorithm.